politeiad/backendv2/tstorebe/plugins/pi/cmds.go

```diff
@@ -0,0 +1,283 @@
+// license that can be found in the LICENSE file.
+
+package pi
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/backendv2/tstorebe/store"
+	"github.com/decred/politeia/politeiad/plugins/pi"
+	"github.com/decred/politeia/politeiad/plugins/ticketvote"
+	"github.com/decred/politeia/util"
+)
+
+const (
+	pluginID = pi.PluginID
+
+	// Blob entry data descriptors
+	dataDescriptorBillingStatus = pluginID + "-billingstatus-v1"
+)
+
+// cmdBillingStatus sets proposal's billing status.
+func (p *piPlugin) cmdBillingStatus(token []byte, payload string) (string, error) {
+	// Decode payload
+	var sbs pi.SetBillingStatus
+	err := json.Unmarshal([]byte(payload), &sbs)
+	if err != nil {
+		return "", err
+	}
+
+	// Verify token
+	err = tokenMatches(token, sbs.Token)
+	if err != nil {
+		return "", err
+	}
+
+	// Verify billing status
+	switch sbs.Status {
+	case pi.BillingStatusClosed, pi.BillingStatusCompleted:
+		// These are allowed; continue
+
+	case pi.BillingStatusActive:
+		// We don't currently allow the status to be manually set to
+		// active.
+		return "", backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeBillingStatusChangeNotAllowed),
+			ErrorContext: "cannot set to active",
+		}
+
+	default:
+		// Billing status is invalid
+		return "", backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeBillingStatusInvalid),
+			ErrorContext: "invalid billing status",
+		}
+	}
+
+	// Verify signature
+	msg := sbs.Token + strconv.FormatUint(uint64(sbs.Status), 10) + sbs.Reason
+	err = util.VerifySignature(sbs.Signature, sbs.PublicKey, msg)
+	if err != nil {
+		return "", convertSignatureError(err)
+	}
+
+	// Ensure reason is provided when status is set to closed.
+	if sbs.Status == pi.BillingStatusClosed && sbs.Reason == "" {
+		return "", backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeBillingStatusChangeNotAllowed),
+			ErrorContext: "must provide a reason when setting " +
+				"billing status to closed",
+		}
+	}
+
+	// Ensure no billing status already exists
+	statuses, err := p.billingStatuses(token)
+	if err != nil {
+		return "", err
+	}
+	if len(statuses) > 0 {
+		return "", backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeBillingStatusChangeNotAllowed),
+			ErrorContext: "can not set billing status more than once",
+		}
+	}
+
+	// Ensure record's vote ended and it was approved
+	vsr, err := p.voteSummary(token)
+	if err != nil {
+		return "", err
+	}
+	if vsr.Status != ticketvote.VoteStatusApproved {
+		return "", backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeBillingStatusChangeNotAllowed),
+			ErrorContext: "setting billing status is allowed only if " +
+				"record was approved",
```

Comment:
```suggestion
				"proposal vote was approved",
```


"record approved" could be interpreted as the record being made public.